File: math/countPrimes.js
// Given an integer n, return the number of prime numbers that are strictly less than n.
/**
 * @param {number} n
 * @return {number}
 */

/*
Use sieve of eratosthenes to fill an array of size n+1 with boolean values

Time complexity: O(n log(log n))
Space complexity: O(n)
*/
var countPrimes = function(n) {
    // Creating the array of booleans to track if a number is prime. 
    // Number is referenced by the index of the array
    let result = 0;
    const arr = new Array(n).fill(true);
    arr[0] = false;
    arr[1] = false;

    // sieve of eratosthenes
    for (let i = 2; i*i < n; i++) {
        if (!arr[i]) continue;
        for (let j = i*i; j<n; j+=i) {
            arr[j] = false;
        }
    }

    // count the number of boolean = true
    for (let a = 0; a < arr.length; a++) {
        if (arr[a]) result += 1;
    }
    
    return result;
};

File: math/count_primes.py
def SieveOfEratosthenes(num):
    prime = [True for i in range(num+1)]
    # boolean array[True,True,True,True,True,True]
    p = 2
    while (p * p <= num):
 
        # If prime[p] is not
        # changed, then it is a prime
        if (prime[p] == True):
            # Updating all multiples of p [True,True,True,True,False,True]
            for i in range(p * p, num+1, p): 
                prime[i] = False
        p += 1
    count = 0
    # count all prime numbers
    for p in range(2, num+1):
        if prime[p]:
            count += 1
    
    return count
